### PR TITLE
Preparing upgrade of analyzers

### PR DESCRIPTION
**Pull request by jnyrup**
## IMPORTANT 

* [ ] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [ ] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [ ] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [ ] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [ ] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

I'm working on migrating our code from the now deprecated `Microsoft.CodeAnalysis.FxCopAnalyzers` and into [.NET analyzers](https://learn.microsoft.com/en-gb/visualstudio/code-quality/migrate-from-fxcop-analyzers-to-net-analyzers?view=vs-2019).

This PR focuses on the lowest hanging fruit as the remaining ones might require a discussion about whether we should disable the diagnostics globally or on a case by case basis.

### CODE REVIEW

File: Src/FluentAssertions/Events/EventRecorder.cs [hunk 1]
```diff
@@ -65,10 +65,10 @@ public void Attach(WeakReference subject, EventInfo eventInfo)
 
     public void Dispose()
     {
-        if (cleanup is not null)
+        if (this.cleanup is { } cleanup)
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Events/EventRecorder.cs:68]: I actually think this makes it _less_ readable.

File: Src/FluentAssertions/Equivalency/Steps/DataRowEquivalencyStep.cs [hunk 2]
```diff
@@ -205,15 +205,34 @@ private class SelectedDataRowMembers
                     new MemberSelectionContext(comparands.CompileTimeType, comparands.RuntimeType, config));
             }
 
-            selectedMembers = new SelectedDataRowMembers
-            {
-                HasErrors = members.Any(m => m.Name == nameof(DataRow.HasErrors)),
-                RowState = members.Any(m => m.Name == nameof(DataRow.RowState))
-            };
+            selectedMembers = GetSelectedDataRowMembers(members);
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Equivalency/Steps/DataRowEquivalencyStep.cs:208]: Wouldn't a simple `ToArray()` not be enough?

Comment (jnyrup) [reply Src/FluentAssertions/Equivalency/Steps/DataRowEquivalencyStep.cs:208]: That forces a _complete_ enumeration of `members` where as this uses _at most_ a single enumeration, but I agree with your preference in using `ToArray`.

File: Src/FluentAssertions/Events/EventRecorder.cs (hunk 1 above)
Comment (jnyrup) [reply Src/FluentAssertions/Events/EventRecorder.cs:68]: Would you prefer something like this?

```cs
Action localCleanup = cleanup;
if (localCleanup is not null)
{
    localCleanup();
    cleanup = null;
    EventObject = null;
    raisedEvents.Dispose();
}
```

Comment (dennisdoomen) [reply Src/FluentAssertions/Events/EventRecorder.cs:68]: Yeah, even though that's ugly too, it's _less_ difficult to read than comparing with `{ }`

File: Src/FluentAssertions/Equivalency/Steps/DataRowEquivalencyStep.cs (hunk 2 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Equivalency/Steps/DataRowEquivalencyStep.cs:208]: I reckon that's negligible with the kinds of assertions people would be doing.

**Pull request merged**